probe-rs/src/architecture/arm/core/m4.rs

```diff
@@ -635,6 +635,33 @@ impl<'probe> CoreInterface for M4<'probe> {
     fn architecture(&self) -> Architecture {
         Architecture::Arm
     }
+
+    /// Read the hardware breakpoints from FpComp registers, and adds them to the Result Vector. A value of 0 in any position of the Vector indicates that the position is unset/available.
+    fn get_hw_breakpoints(&mut self) -> Result<Vec<u32>, Error> {
+        let mut breakpoints = vec![];
+        let num_hw_breakpoints = self.get_available_breakpoint_units()? as usize;
+        { 0..num_hw_breakpoints }.try_for_each(|bp_unit_index| {
+            let raw_val = self.memory.read_word_32(FpCtrl::ADDRESS)?;
+            let ctrl_reg = FpCtrl::from(raw_val);
+            let breakpoint: u32;    // The breakpoint address after it has been adjusted for FpRev
+            let register_value: u32;// The raw breakpoint address as read from memory
+            if ctrl_reg.rev() == 0 {
+                let reg_addr = FpRev1CompX::ADDRESS + (bp_unit_index * size_of::<u32>()) as u32;
+                register_value = self.memory.read_word_32(reg_addr)?;
+                breakpoint = register_value & 0x1f_ff_ff_fc;
+            } else if ctrl_reg.rev() == 1 {
+                let reg_addr = FpRev2CompX::ADDRESS + (bp_unit_index * size_of::<u32>()) as u32;
+                register_value = self.memory.read_word_32(reg_addr)?;
```

Comment:
I think we should also check if the breakpoint unit is actually enabled, to avoid confusion. Each register has a bit 0 which indicates if the HW breakpoint is actually active.

Reply:
@Tiwalun ... If I understand you correctly, you want us to return a 'disabled' breakpoint as a None? In other words the consumer will interpret that they can overwrite it with a new comparator?

Reply:
I have a proposal ... want to bounce it off you to see if it covers all cases. 
1. `get_hw_breakpoints` returns all, irrespective of whether they are enabled.
2. `set_hw_breakpoint` will set it if it doesn't exist in the list above, or enable it if it exists
3. `clear_hw_breakpoint` will clear the breakpoint, irrespective of whether it is enabled or not
4. `clear_all_hw_breakpoints` will clear the breakpoint, also unconditionally

Final note. In the core mode, number 2 and 3 above are qualified with a `_hw_`, but in the architecture specific mods they are just `set_breakpoint` and `clear_breakpoint`. Are you OK if I rename the latter to include the `_hw_` qualifier ... in prep of introducing software breakpoints?

Reply:
@Tiwalun ... I've implemented changes as per my proposal above. Want to try it out and let me know what you think?

Reply:
> @Tiwalun ... If I understand you correctly, you want us to return a 'disabled' breakpoint as a None? In other words the consumer will interpret that they can overwrite it with a new comparator?

Regarding the enabled / disabled state, currently a use can call the `clear_hw_breakpoint`, and this means both setting the address to 0, as well as disabling the breakpoint. A disabled breakpoint with an address should never happen right now, if only the provided breakpoint interfaces are used.

I'm not sure if offering a special interface to disable a breakpoint without changing the address even makes any sense, because for that we first need to read the breakpoint address, and then write back the disabled state. If we just want to disable a breakpoint, just writing all zeros is quicker.



> How true do you want to stay to the DAP Specification? The current spec will ALWAYS clear all breakpoints before setting a new set of breakpoints. If we want to support this with your change, we should

The DAP specification should not influence the API too much, it should also be usable for other purposes.
The debug adapter should keep track of disabled breakpoints separately, if it needs access to disabled breakpoints. The number of disabled breakpoints should not be limited by the number of HW breakpoints, it should be independent.


> Final note. In the core mode, number 2 and 3 above are qualified with a `_hw_`, but in the architecture specific mods they are just `set_breakpoint` and `clear_breakpoint`. Are you OK if I rename the latter to include the `_hw_` qualifier ... in prep of introducing software breakpoints?

Renaming the functions is a good idea.